# The List Data Type


In [ ]:
my_list_var =  [1, 2, 3]

In [68]:
type(my_list_var)

list

In [69]:
your_var = 3
type(your_var)

int

In [70]:
animals = ['cat', 'bat', 'horse', 'elephant']
animals

['cat', 'bat', 'horse', 'elephant']

In [71]:
my_list_with_mixed_data_type= ['hello', 3.1415, True, None, 42]
my_list_with_mixed_data_type

['hello', 3.1415, True, None, 42]

In [72]:
spam = ['cat', 'bat', 'rat', 'elephant']

# Getting Individual Values in a List with Indexes


In [74]:
spam[2]

'rat'

In [ ]:
my_list_with_mixed_data_type[1]

# Negative Indexes


- you can use any valid variable names for naming the list

In [75]:
spam = ['cat', 'bat', 'rat', 'elephant']

In [76]:
spam[-1]

'elephant'

In [78]:
spam[-4]

'cat'

In [79]:
spam[0]

'cat'

In [80]:
index = 0
spam[index]

'cat'

In [ ]:
my_list_with_mixed_data_type[5]

# Getting a List from Another List with Slices
- one integer index
- two integer slice


In [81]:
animals = ['cat', 'bat', 'rat', 'elephant', 'horse']
animals

['cat', 'bat', 'rat', 'elephant', 'horse']

In [88]:
animals[1:50] # slicing the list. first is inclusive


['bat', 'rat', 'elephant', 'horse']

In [92]:
animals[50]

IndexError: list index out of range

In [90]:
animals = ['cat', 'bat', 'rat', 'elephant', 'horse',[1,2,3]]
animals[-1]


[1, 2, 3]

In [93]:
animals = ['cat', 'bat', 'rat', 'elephant', 'horse',[1,2,3]]
animals[[1,2]]

TypeError: list indices must be integers or slices, not list

In [94]:
animals

['cat', 'bat', 'rat', 'elephant', 'horse', [1, 2, 3]]

In [96]:
animals[:3]# from the beginning till inded 4

['cat', 'bat', 'rat']

In [97]:
animals[2:] # from the index 2, till the end

['rat', 'elephant', 'horse', [1, 2, 3]]

In [ ]:
middle_point = len(any_name)//2 # pirovt_index variable to store the middle point of the list
any_name = any_name[middle_point:]
any_name

In [ ]:
animals

In [ ]:
len(animals)

In [ ]:
middle_point = len(animals)//2
middle_point


In [ ]:
animals[middle_point]

# Getting a List’s Length with the len() Function
- depending on the data type meaning of operators can change
    - len()
    - *

In [98]:
len('student')

7

In [100]:
type('student')

str

In [102]:
len(['student', 'perseverance'])

2

In [101]:
type(['student'])

list

In [103]:
2*3

6

In [104]:
'hello'*3

'hellohellohello'

In [105]:
'hello'/3

TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [ ]:
len('message')

In [ ]:
len(any_name)

In [ ]:
any_name

In [ ]:
animals

In [ ]:
len(animals) 

# Changing Values in a List with Indexes


In [106]:
ta_students= ['Asadbek', 'Shamskamar']
ta_students

['Asadbek', 'Shamskamar']

In [107]:
ta_students[0] = 'Humoyun' # to change the value fo the list you can use indexing+assignment operator
ta_students

['Humoyun', 'Shamskamar']

In [109]:
ta_students = ['TA1','TA2']
ta_students

['TA1', 'TA2']

In [110]:
ta_students

['TA1', 'TA2']

In [111]:
ta_students[-1] = 'TA64'
ta_students

['TA1', 'TA64']

# List Concatenation and List Replication


In [113]:
TAs_of_001 = ['Javokhir','Habibullo']
TAs_of_002 = ['Asadbek','Shamsqamar']



In [112]:
'Hello' +' '+'Class'

'Hello Class'

In [ ]:
s = 'Hello' +' '+'Class' # this is a concatation expression
print(s)

In [ ]:
len(s)

In [114]:
TAs = TAs_of_001 + TAs_of_002  # concatematopm operator 
TAs

['Javokhir', 'Habibullo', 'Asadbek', 'Shamsqamar']

In [115]:
animals

['cat', 'bat', 'rat', 'elephant', 'horse', [1, 2, 3]]

In [118]:
second_animals  =['bird']
animals = second_animals+['fish']
animals

['bird', 'fish']

In [ ]:
TAs

In [119]:
students_score = [ 75 ] #62 stuydents 
students_score

[75]

In [120]:
len(students_score)

1

In [121]:
# make stuents_score variable store every students score: average
students_score = [ 70] * 62
len(students_score)

62

In [122]:
students_score

[70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70,
 70]

In [123]:
import random
#Generate 5 random numbers between 10 and 30
randomlist = random.sample(['1','2','4','5'], 2)
print(randomlist)

['4', '1']


# Removing Values from Lists with del Statements


In [133]:
contacts = ["Person A", "Person B", "Person X"] # image 1000x1000 pixes 10^6
contacts

['Person A', 'Person B', 'Person X']

In [137]:
del contacts[-1]
contacts

IndexError: list assignment index out of range

In [129]:
contacts

['Person X']

In [ ]:
contacts[1:]

In [130]:
del contacts


In [131]:
contacts

NameError: name 'contacts' is not defined

In [ ]:
del contacts[:2]
contacts

In [ ]:
del contacts

In [ ]:
contacts

In [ ]:
del contacts[2] 
contacts

In [ ]:
del contacts[0] 
contacts

# Working with Lists
- lists can help store a group of similar values in a better way
    - more organized
    - less duplicate

In [ ]:
catName1 = 'Zophie'
catName2 = 'Pooka'
catName3 = 'Simon'
catName4 = 'Lady Macbeth'
catName5 = 'Fat-tail'
catName6 = 'Miss Cleo'

In [ ]:
# see the example codes

# For loops with lists

In [138]:
supplies = ['pens', 'staplers', 'flamethrowers', 'binders']

for i in range(0,len(supplies)):
    print('item with index',i, 'is',supplies[i])

item with index 0 is pens
item with index 1 is staplers
item with index 2 is flamethrowers
item with index 3 is binders


In [139]:
for item in supplies:
    print("item at index ", item)

item at index  pens
item at index  staplers
item at index  flamethrowers
item at index  binders


# Using Enumerate

In [144]:

for any_variable_name,item in enumerate(supplies):
    print("item at index ",any_variable_name, 'is', item)

item at index  0 is pens
item at index  1 is staplers
item at index  2 is flamethrowers
item at index  3 is binders


# The in and not in Operators
- in checks if element is inside the list
- not in checks if the element is not inside the list

In [141]:
'howdy' in ['hello', 'hi', 'howdy', 'heyas']

True

In [142]:
'howdy' not in ['hello', 'hi', 'howdy', 'heyas']

False

In [143]:
my_list = ['hello', 'hi', 'howdy', 'heyas']
'howdy' in my_list

True

In [145]:
my_list.index('howdy')

2

In [146]:
for i, item in enumerate(my_list):
    if item=='howdy':
        I = i
        break
    

In [147]:
I

2

# Using the random.choice() and random.shuffle() Functions with Lists
- choice will randomly select index
- shuffle shuffles the order

In [149]:
import random
my_list

['hello', 'hi', 'howdy', 'heyas']

In [150]:

random.choice(my_list)

'howdy'

In [151]:

random.choice(my_list)

'hello'

In [161]:
selections = ["Lucky","Not so much"]

answer = random.choice(selections)
if answer =='Lucky':
    print("You won")
else:
    print("Try again")

You won


In [162]:
my_list = ['hello', 'hi', 'howdy', 'heyas']
random.shuffle(my_list)
my_list

['hi', 'howdy', 'heyas', 'hello']

In [ ]:
my_list

#  Augmented Assignment Operators
- When assigning a value to a variable, we frequently use the variable itself. 
- augmented operators
    - `+`
    - `-`
    - `*`
    - `/`
    - `%`

In [164]:
age = 10 # assing


In [165]:
age +=1 # increase variable age by 1, then assign
age

11

In [166]:
age = age+1
age

12

In [168]:
age = 10
age=age%7
age

3

# Methods


## finding a value in a list with the index() Method


In [170]:
spam = ['hello', 'hi', 'howdy', 'heyas']
spam.index('not to be found')

ValueError: 'not to be found' is not in list

## Adding Values to Lists with the append() and insert() Methods


In [171]:
spam = ['cat', 'dog', 'bat']
spam.append("elephant")
spam


['cat', 'dog', 'bat', 'elephant']

In [ ]:
spam

In [172]:
spam.append("horse")

In [173]:
spam

['cat', 'dog', 'bat', 'elephant', 'horse']

In [174]:
spam

['cat', 'dog', 'bat', 'elephant', 'horse']

In [175]:
spam.insert(3,'lion')
spam

['cat', 'dog', 'bat', 'lion', 'elephant', 'horse']

In [176]:
spam.insert('lion')
spam

TypeError: insert expected 2 arguments, got 1

In [ ]:
spam

In [ ]:
# put elephant in the begninnig
del spam[4]
spam 

In [ ]:
spam.insert(0,'elaphant')
spam

In [ ]:
# 
spam = ['cat', 'dog', 'bat']
spam.append("elephant")
spam

In [ ]:
spam[0]='elaphant'
spam[3]='cat'
spam

In [ ]:
# 
spam = ['cat', 'dog', 'bat']
spam.append("elephant")
spam

In [ ]:
spam.insert(0,spam.pop(3))
spam

## Removing Values from Lists with the remove() Method
 

In [177]:
spam = ['cat', 'bat', 'rat', 'elephant']


In [178]:
del spam[1]
spam

['cat', 'rat', 'elephant']

In [179]:
spam = ['cat', 'bat', 'rat', 'elephant']
spam


['cat', 'bat', 'rat', 'elephant']

In [180]:
spam.remove('bat')
spam


['cat', 'rat', 'elephant']

### del vs remove
- if u have index, then use del
- if u have the name then remove

## Sorting the Values in a List with the sort() Method


In [181]:
ages = [10,20,15,30]
ages

[10, 20, 15, 30]

In [182]:
type(ages)

list

In [183]:
ages.sort()

In [184]:
ages

[10, 15, 20, 30]

In [186]:
ages.sort(reverse=True)

In [187]:
ages

[30, 20, 15, 10]

In [ ]:
ages

In [ ]:
ages.sort()
ages

In [ ]:
ages.reverse()

In [ ]:
ages

In [ ]:
ages.sort(reverse=True)
ages

## Reversing the Values in a List with the reverse() Method


## Example Program 
- write a code for randomly selecting a student from the list of students


In [ ]:
div002_student_names = [
    'Mukhammadov Muminjon Otajon Ugli',
    'Kutlimuratov Madrakhim Khushnud Ugli',
    'Akimov Sarvar Ravshan Ugli',
    'Akhmadjonov Samandar Anvarjon Ugli',
    'Shokirov Saidalo Nematillo Ugli',
    'Sharifov Azizbek Sharofovich',
    'Khudaynazarov Sunnatbek Bakhodir Ugli',
    'Carrasco Soriano Diana Laura',
    'Zokirov Diyorbek Botir Ugli',
    'Muhammadkhonov Mavlonkhon Muhriddinkho',
    'Vu Nguyen Bao Ngoc',
    'Murtozaev Sardor Muzaffar Ugli',
    'Rustamov Shamsiddin Azam Ugli',
    'Kakharova Sevinch Javlon Kizi',
    'Jabborov Sunnatjon Bakhtiyor Ugli',
    'Hamidkhonov Amirkhon Obidkhon Ugli',
    'Mustafakulov Azizbek Furkat Ugli',
    'Ziyaviddinov Ibrohim Ilhom Ugli',
    'Turaev Jahongir Bakhtiyorovich',
    'Ishonkulov Jakhongir Khusan Ugli',
    'Chinieva Shokhida Dilshodovna',
    'Khamdamov Kamronbek Laziz Ugli',
    'Tursunov Firdavs Farhodovich',
    'Ha Thi My Linh',
    'Kurbonova Nilufar Mizrobovna',
    'Orziev Diyorbek Matnazar Ugli',
    'Obidov Dilshod Khurshedovich',
    'Togaev Ravshan Khayritdin Ugli',
    'Shodmonov Rustam Bakhtiyor Ugli',
    'Khudoyberdiev Madaminbek Pulat Ugli',
    'Ilhomov Mansur Ilhom Ugli',
    'Kakhorjonov Mansurbek Tokhirjon Ugli',
    'Tsogoo Munkhzul',
    'Adhamov Muhammadaziz Gayratjon Ugli',
    'Koilybay Miras',
    'Muhammedov Mironshoh Mirkhon Ugli',
    'Abdurasulov Mirsaid Alimardonovich',
    'Nematjonov Bahodir Nurmukhamad Ugli',
    'Pak Georgiy',
    'Abduvokhidov Bekmurod Abdulkhodi Ugli',
    'Eshmirzaev Sarvar Raufjonovich',
    'Kosimov Shamskamar Jahongir Ugli',
    'Baratov Sokhibjon Khurshedovich',
    'Shokulov Shohruh Tuymurod Ugli',
    'Meliboev Shukurullo Fakhriddin Ugli',
    'Baktybek Uulu Arsen',
    'Shermirzaev Abbosbek Isroiljon Ugli',
    'Kiyomov Asadbek Kosimjon Ugli',
    'Mukhtorjonov Asilbek Murodjon Ugli',
    'Suiindik Akniyet',
    'Nabijonov Javokhir Pulatjon Ugli',
    'Nimatilaev Javokhir Izatillo Ugli',
    'Khursanaliev Javokhir Mukhiddin Ugli',
    'Abdulkhodiev Jobirkhon Rus Tamjon Ugli',
    'Malabon Julliane Samantha',
    'Chaw Thiri San',
    'Sobitkhonov Kudratkhon Odilkhon Ugli',
    'Bekzhanov Tariel Bekzhanovich',
    'Raykhonov Tilov Alisher Ugli',
    'Otamirzaev Fayozbek Farkhodjon Ugli',
    'Pirmatov Feruzbek Ermat Ugli',
    'Davlatov Hamidullo Nosirjon Ugli'
 ]

# autumation example

In [191]:
import pandas as pd
df = pd.read_excel('div_002.xlsx')
df

No.   ID number                              Fullname Department
0    NaN         NaN                                   NaN        NaN
1   62.0  12214687.0      Mukhammadov Muminjon Otajon Ugli      IBT학과
2   61.0  12214732.0  Kutlimuratov Madrakhim Khushnud Ugli      ISE학과
3   60.0  12214739.0            Akimov Sarvar Ravshan Ugli      ISE학과
4   59.0  12214740.0    Akhmadjonov Samandar Anvarjon Ugli      ISE학과
..   ...         ...                                   ...        ...
58   5.0  12225274.0         Bekzhanov Tariel Bekzhanovich      ISE학과
59   4.0  12225275.0          Raykhonov Tilov Alisher Ugli      ISE학과
60   3.0  12225276.0   Otamirzaev Fayozbek Farkhodjon Ugli      ISE학과
61   2.0  12225277.0          Pirmatov Feruzbek Ermat Ugli      ISE학과
62   1.0  12225278.0      Davlatov Hamidullo Nosirjon Ugli      ISE학과

[63 rows x 4 columns]

In [194]:
df['Fullname'].values.tolist()

[nan,
 'Mukhammadov Muminjon Otajon Ugli',
 'Kutlimuratov Madrakhim Khushnud Ugli',
 'Akimov Sarvar Ravshan Ugli',
 'Akhmadjonov Samandar Anvarjon Ugli',
 'Shokirov Saidalo Nematillo Ugli',
 'Sharifov Azizbek Sharofovich',
 'Khudaynazarov Sunnatbek Bakhodir Ugli',
 'Carrasco Soriano Diana Laura',
 'Zokirov Diyorbek Botir Ugli',
 'Muhammadkhonov Mavlonkhon Muhriddinkho',
 'Vu Nguyen Bao Ngoc',
 'Murtozaev Sardor Muzaffar Ugli',
 'Rustamov Shamsiddin Azam Ugli',
 'Kakharova Sevinch Javlon Kizi',
 'Jabborov Sunnatjon Bakhtiyor Ugli',
 'Hamidkhonov Amirkhon Obidkhon Ugli',
 'Mustafakulov Azizbek Furkat Ugli',
 'Ziyaviddinov Ibrohim Ilhom Ugli',
 'Turaev Jahongir Bakhtiyorovich',
 'Ishonkulov Jakhongir Khusan Ugli',
 'Chinieva Shokhida Dilshodovna',
 'Khamdamov Kamronbek Laziz Ugli',
 'Tursunov Firdavs Farhodovich',
 'Ha Thi My Linh',
 'Kurbonova Nilufar Mizrobovna',
 'Orziev Diyorbek Matnazar Ugli',
 'Obidov Dilshod Khurshedovich',
 'Togaev Ravshan Khayritdin Ugli',
 'Shodmonov Rustam Bakh

# The Tuple Data Type
- mutable vs immutable
    - list is mutable (can be changed)
    - tuple is immutable (cannot be modifed)

In [200]:
my_list = (1,2,3,4)
my_list[0]=-5

TypeError: 'tuple' object does not support item assignment

In [196]:
my_list[0]=-10
my_list

[-10, 2, 3, 4]

In [197]:
my_tuple = (2,3,4,5)
my_tuple

(2, 3, 4, 5)

In [198]:
my_tuple[0]

2

In [199]:
my_tuple[0] = -2

TypeError: 'tuple' object does not support item assignment

# Converting Types with the list() and tuple() Functions


In [201]:
my_tuple = ('apple','orange','peach')
my_tuple

('apple', 'orange', 'peach')

In [202]:
my_list = list(my_tuple)
my_list

['apple', 'orange', 'peach']

In [203]:
tuple_var = tuple(my_list)
tuple_var

('apple', 'orange', 'peach')

## References


In [210]:
# simple data types
def increment_age(age):
    age+=1
    
age = 10
increment_age(age)

In [211]:
age

10

In [208]:
# simple data types
def increment_age_list(age_list):
    age_list[0] = 100
    
age_list = [10,20,30,40]
increment_age_list(age_list)

In [209]:
age_list

[100, 20, 30, 40]

In [ ]:
# lists
spam = [0, 1, 2, 3, 4, 5]
cheese = spam # The reference is being copied, not the list.
cheese[1] = 'Hello!' # This changes the list value.

In [ ]:
ag

## Identity and the id() Function


In [212]:
age 

10

In [213]:
id('Howdy') # The returned number will be different on your machine.

2557475440368

In [214]:
bacon = 'Hello'
id(bacon)

2557544586800

In [215]:
id(bacon)

2557544586800

In [216]:
bacon += ' world!' # A new string is made from 'Hello' and ' world!'.
id(bacon)

2557475729200

## Passing references
- list names are passed as a reference

In [217]:
def eggs(someParameter):
    someParameter.append('Hello')

spam = [1, 2, 3]
eggs(spam)
spam

[1, 2, 3, 'Hello']

- how to prevent that?

## The copy Module’s copy()  Functions


In [218]:
import copy
def eggs(someParameter):
    someParameter.append('Hello')

spam = [1, 2, 3]
spam2 = spam
eggs(spam2)
spam

[1, 2, 3, 'Hello']

In [221]:
import copy as cp # is a module: copy and deepcopy
def eggs(someParameter):
    someParameter.append('Hello')

spam = [1, 2, 3]
spam2 = cp.copy(spam)
eggs(spam2)
spam

[1, 2, 3]

In [222]:
import copy as cp # is a module: copy and deepcopy
def eggs(someParameter):
    someParameter.append('Hello')

spam = [1, 2, 3, ['a','b','c']]
spam2 = cp.copy(spam)
eggs(spam2)
spam

[1, 2, 3, ['a', 'b', 'c']]

In [220]:
spam2

[1, 2, 3, 'Hello']

# Homework 5

### Q0. for Q0, you write a short program
- define a function that 
    - takes two params as shown below
    - removes given subject from the list
    - returns the reduced list
- run the function definition
- call your defined function with the following arguments


In [ ]:
def drop_subject(all_subjects, name_of_the_subject_to_drop):
    # --write your code below this line--

    # remove the subject to be dropped from the list
    # return reduced list
    


In [ ]:
### Then call your defined function with the following arguments:
drop_subject(['SW','Korean Lang','Math','Business','Other'], 'SW')


### Q1. What is []?

### Q2. How would you assign the value 'hello' as the third value in a list stored in a variable named spam? (Assume spam contains [2, 4, 6, 8, 10].)




For the following three questions, let’s say spam contains the list ['a', 'b', 'c', 'd'].

### Q3. What does spam[int(int('3' * 2) // 11)] evaluate to?

### Q4. What does spam[-1] evaluate to?

### Q5. What does spam[:2] evaluate to?


For the following three questions, let’s say bacon contains the list [3.14, 'cat', 11, 'cat', True].

## Q6. What does bacon.index('cat') evaluate to?

## Q7. What does bacon.append(99) make the list value in bacon look like?

## Q8. What does bacon.remove('cat') make the list value in bacon look like?